### PR TITLE
feat: datatypes for LSP code actions

### PR DESCRIPTION
**Pull request by EdAyers**
A subset of https://github.com/leanprover/lean4/pull/1551 with just the new datatypes for the LSP protocol. I have also added documentation (copied from spec) and reference links to the LSP specification.

### CODE REVIEW

File: src/Lean/Data/Lsp/Basic.lean [hunk 1]
```diff
@@ -91,20 +123,141 @@ structure VersionedTextDocumentIdentifier where
...
+  uri           : DocumentUri
+  options?      : Option DeleteFile.Options := none
+  annotationId? : Option String := none
+  deriving ToJson, FromJson
 
+/-- A change to a file resource.
+
+[reference](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#resourceChanges) -/
+inductive DocumentChange where
+  | create : CreateFile       → DocumentChange
```
Comment (EdAyers) [inline src/Lean/Data/Lsp/Basic.lean:182]: Can I get this to work with `derive ToJson`?

File: src/Lean/Data/Json/FromToJson.lean [hunk 2]
```diff
@@ -133,6 +133,14 @@ instance : FromJson Float where
     | (Json.num jn) => Except.ok jn.toFloat
     | _ => Except.error "Expected a number or a string 'Infinity', '-Infinity', 'NaN'."
 
+instance {cmp} [ToJson α]: ToJson (RBMap String α cmp) where
```
Comment (Kha) [inline src/Lean/Data/Json/FromToJson.lean:136]:
```suggestion
instance [ToJson α] : ToJson (RBMap String α cmp) where
```
`{cmp}` shouldn't do anything here, does it?

File: src/Lean/Data/Lsp/Basic.lean [hunk 3]
```diff
@@ -82,6 +106,14 @@ instance : FromJson TextEditBatch :=
 instance : ToJson TextEditBatch :=
   ⟨@toJson (Array TextEdit) _⟩
 
+instance : EmptyCollection TextEditBatch := ⟨#[]⟩
+
+instance : Append TextEditBatch :=
+  show Append (Array TextEdit) by infer_instance
```
Comment (Kha) [inline src/Lean/Data/Lsp/Basic.lean:112]:
```suggestion
  inferInstanceAs (Append (Array _))
```

**Review by Kha — APPROVED** [review]

File: src/Lean/Data/Lsp/Basic.lean (hunk 1 above)
Comment (Kha) [reply src/Lean/Data/Lsp/Basic.lean:182]: Not right now I guess

**Pull request merged**